compat/mswindows.cc

```diff
@@ -203,7 +204,7 @@ _free_osfhnd(int filehandle)
             SetStdHandle(STD_ERROR_HANDLE, nullptr);
             break;
         }
-        _osfhnd(filehandle) = (long) INVALID_HANDLE_VALUE;
+        _osfhnd(filehandle) = reinterpret_cast<intptr_t>(INVALID_HANDLE_VALUE);
```

Comment:
This cast does not make sense to me (in Squid context) because Squid defines _osfhnd() to return a `long` data member reference:

```C++
/* internal to Microsoft CRTLIB */
typedef struct {
    long osfhnd;    /* underlying OS file HANDLE */
...
}   ioinfo;
```

The above struct is used by the _osfhnd() macro used by the PR-changed code:

```C++
#define _osfhnd(i)  ( _pioinfo(i)->osfhnd )
#define _pioinfo(i) ( __pioinfo[...] + ... )
SQUIDCEXTERN _CRTIMP ioinfo * __pioinfo[];
```

AFAICT, [some](https://doxygen.reactos.org/d7/d45/crtdefs_8h.html#a0fbe4a4f8dd857ee04923a901f27465f) code defines that `osfhnd` member as `intptr_t` so I am guessing it is the modern way to define that data member, but the above PR change is incompatible with our own `ioinfo` definition...

I do not know whether our own definition should be removed (e.g., as no longer unused) or this comparison code should be made smarter about the actual type of the `osfhnd` data member (to accommodate cases where our own definition _is_ used).

Reply:
What I think Squid is doing in that context is to define a copy of the Windows internals to keep the compiler happy in some contexts.

Here is what the preprocessed code looks like:

```
    if (((unsigned)filehandle < 2048) &&
        (( ( __pioinfo[(filehandle) >> 5] + ((filehandle) & ((1 << 5) - 1)) )->osfile ) & 0x01) &&
        (( ( __pioinfo[(filehandle) >> 5] + ((filehandle) & ((1 << 5) - 1)) )->osfhnd ) != reinterpret_cast<intptr_t>( ((HANDLE) (LONG_PTR)-1) )))
```
http://www.shrutisagarashram.org/ashram/VC/VC98/CRT/SRC/OSFINFO.C looks a lot like the code we are using; maybe it served as inspiration

I suspect we will want to revisit Windows IO, but we need to get to compile first

Reply:
Your comment pretty much matches my understanding and the reasons behind my change request. Yes, we are defining what some Windows flavors are probably using internally, but this PR changes are not compatible with our (old) definitions. Secondary problems aside, on the surface, it looks like this PR fixes one use case by breaking another. We should not do that, especially when that "other use case" is based on our own code (that defines `ioinfo`).

> Here is what the preprocessed code looks like: ...

Is that based on Squid-provided `ioinfo` definition where `osfhnd` is `long`? Or is it based on Windows-provided `ioinfo` definition? If it is the latter, what is `osfhnd` type? As I have mentioned in the change request, I have seen that type being a pointer (rather than `long`) in some environments, which could explain why your changes fix the build (in those environments), but there are other explanations as well.

Your should be able to answer the first question by, for example, commenting out our `ioinfo` definition or renaming its  `osfhnd` field (and adjusting the corresponding accessor macro). If the preprocessed code does not change, then you are not using our definition and, hence, is not reproducing the PR problem that will surface in environments that do use that definition.